source/state_representation/src/trajectory/CartesianTrajectory.cpp

```diff
@@ -0,0 +1,210 @@
+    );
+  }
+  CartesianTrajectoryPoint trajectory_point;
+  trajectory_point.data = new_point.data();
+  trajectory_point.duration = duration;
+  trajectory_point.name = new_point.get_name();
+  try {
+    this->TrajectoryBase<CartesianTrajectoryPoint>::insert_point(trajectory_point, index);
+  } catch (...) {
+    throw;
+  }
+}
+
+void CartesianTrajectory::set_point(
+    const CartesianState& point, const std::chrono::nanoseconds& duration, unsigned int index
+) {
+  if (point.is_empty()) {
+    throw exceptions::EmptyStateException("Point is empty");
+  } else if (point.get_reference_frame() != this->reference_frame_) {
+    throw exceptions::IncompatibleReferenceFramesException(
+        "Incompatible reference frames: " + point.get_reference_frame() + " and " + this->reference_frame_
+    );
+  }
+  CartesianTrajectoryPoint trajectory_point;
+  trajectory_point.data = point.data();
+  trajectory_point.duration = duration;
+  trajectory_point.name = point.get_name();
+  try {
+    this->TrajectoryBase<CartesianTrajectoryPoint>::set_point(trajectory_point, index);
+  } catch (...) {
+    throw;
+  }
+}
+
+void CartesianTrajectory::set_points(
+    const std::vector<CartesianState>& points, const std::vector<std::chrono::nanoseconds>& durations
+) {
+  if (points.empty()) {
+    throw exceptions::IncompatibleSizeException("No points provided");
+  } else if (points.size() != this->get_size()) {
+    throw exceptions::IncompatibleSizeException("The size of the current vector and the new vector are not equal");
+  } else if (points.size() != durations.size()) {
+    throw exceptions::IncompatibleSizeException("The size of the points and durations vectors are not equal");
+  }
+
+  std::vector<CartesianTrajectoryPoint> trajectory_points;
+  for (unsigned int i = 0; i < points.size(); ++i) {
+    if (points[i].is_empty()) {
+      throw exceptions::EmptyStateException("Vector contains at least one point that is empty");
+    } else if (points[i].get_reference_frame() != this->reference_frame_) {
+      throw exceptions::IncompatibleReferenceFramesException(
+          "Incompatible reference frames: " + points[i].get_reference_frame() + " and " + this->reference_frame_
+      );
+    }
+    CartesianTrajectoryPoint trajectory_point;
+    trajectory_point.data = points[i].data();
+    trajectory_point.duration = durations[i];
+    trajectory_point.name = points[i].get_name();
+    trajectory_points.push_back(trajectory_point);
+  }
+  try {
+    this->TrajectoryBase<CartesianTrajectoryPoint>::set_points(trajectory_points);
+  } catch (...) {
+    throw;
+  }
+}
+
+const std::vector<CartesianState> CartesianTrajectory::get_points() const {
+  std::vector<CartesianState> points;
+  for (unsigned int i = 0; i < this->get_size(); ++i) {
+    points.push_back(this->operator[](i).first);
+  }
+  return points;
+}
+
+CartesianState CartesianTrajectory::get_point(unsigned int index) const {
```

Comment:
Seeing this, I wonder if we miss getters in the base class for `get_duration(unsigned int index)` and `get_time_from_start(unsigned int index)` ?

Reply:
yeah, I think this might be a nice addition for consistency